chat-client.ts

```diff
@@ -1,144 +1,163 @@
-import memoizeOne from "memoize-one";
+import { ChatClientOptions } from "./types";
 
-import { ClientOptions } from "./types";
-
-const initializeWSClient = memoizeOne(async (projectId: string) => {
-  return new Promise<WebSocket>((resolve, reject) => {
-    const wsClient = new WebSocket("wss://api.commonbase.com/chats");
+async function startWebSocketSession(options: ChatClientOptions) {
+  return new Promise<[WebSocket, string]>((resolve, reject) => {
+    const ws = new WebSocket(
+      "wss://api.commonbase.com/chats" +
+        ("sessionId" in options && options.sessionId
+          ? "/" + options.sessionId
+          : ""),
+    );
 
     function cleanup() {
-      wsClient.removeEventListener("error", handleError);
-      wsClient.removeEventListener("message", handleMessage);
+      ws.removeEventListener("error", handleError);
+      ws.removeEventListener("message", handleMessage);
     }
 
-    const handleError = (error: Event) => {
+    const handleError = (error: unknown) => {
       reject(error);
       cleanup();
     };
-    wsClient.addEventListener("error", handleError);
+    ws.addEventListener("error", handleError);
 
     const handleMessage = (event: MessageEvent<string>): void => {
       const msg = JSON.parse(event.data);
-      if (msg.type === "init") {
+      if (msg.type === "init" || msg.type == "info") {
         if (msg.error) {
-          reject(msg.error);
-          cleanup();
+          handleError(msg.error);
           return;
         }
         if (msg.initialized) {
-          if (msg.error) {
+          if (msg.error && msg.initialized !== 0) {
             reject(msg.error);
           } else {
-            resolve(wsClient);
+            const sessionId =
+              msg.sessionId ?? ("sessionId" in options && options.sessionId);
+            if (sessionId) {
+              resolve([ws, sessionId]);
+            } else {
+              reject("Missing sessionId");
+            }
           }
           cleanup();
         }
       }
     };
-    wsClient.addEventListener("message", handleMessage);
+    ws.addEventListener("message", handleMessage);
 
     function handleOpen() {
-      wsClient.send(JSON.stringify({ type: "init", projectId }));
-      wsClient.removeEventListener("open", handleOpen);
+      const message =
+        "sessionId" in options && options.sessionId
```

Comment:
This is completely unreadable. Let's please stick to if's or extract it into a function.

Reply:
I think that's an opinion, but I feel less strongly about it. Replaced it with an if-else which I find a tad less readable